File: src/javascript/algorithms/matrix/4.wordSearch/wordSearch.js
 // DFS Backtracking:- TC: O(m * n * 4^L) SC:O(L)

 function isWordExist(board, word) {
    let rows = board.length, cols = board[0].length;

    for(let r=0; r<rows; r++) {
        for(let c=0; c<cols; c++) {
            //Start dfs with first letter in the board
            return dfs(board, word, r, c, 0)) {
                return true;
            }
        }
    }
    //Not found even after exhaustive search in the board
    return false;
 }

 function dfs(board, word, r, c, curr) {
     // If the entire word has been found in the board
    if(word.length === curr) return true;

    // Verify out of bounds or mismatched letter
    if(r <0 || r>= board.length || c<0 || c>= board[0].length || board[r][c] !== word[curr] || board[r][c] === '*') return false;

    //Mark the current word as visited
    let temp = board[r][c];
    board[r][c] = '*';

    //Explore neighbour letter of current word for the next letter in a word
    let top = dfs(board, word, r-1, c, curr+1);
    let left = dfs(board, word, r, c-1, curr+1);
    let bottom = dfs(board, word, r+1, c, curr+1);
    let right = dfs(board, word, r, c+1, curr+1);

    //Restore the original letter
    board[r][c] = temp;
    return top || left || bottom || right;
 }

let board1 = [
    ['A','B','C','D'], ['W','M','G','O'], ['B','D','C','N']
];
let board2 = [
    ['A','B','C','D'], ['w','A','G','O'], ['B','T','C','N']
];

let word1 = "DOG";
let word2 = "CAT";
console.log(isWordExist(board1, word1));
console.log(isWordExist(board2, word2));

File: src/python/algorithms/matrix/4.wordSearch/word_search.py
"""
Word Search

Given an m x n grid of characters board and a string word, return true if word exists in the grid.
The word can be constructed from letters of sequentially adjacent cells.

Time Complexity: O(m * n * 4^L) where L is word length
Space Complexity: O(L) for recursion stack
"""


def exist(board, word):
    """
    DFS/Backtracking approach.
    TC: O(m * n * 4^L), SC: O(L)
    """
    if not board:
        return False
    
    rows, cols = len(board), len(board[0])
    
    def dfs(r, c, index):
        if index == len(word):
            return True
        
        if (r < 0 or r >= rows or c < 0 or c >= cols or 
            board[r][c] != word[index]):
            return False
        
        # Mark as visited
        temp = board[r][c]
        board[r][c] = '#'
        
        # Explore all 4 directions
        found = (dfs(r + 1, c, index + 1) or
                 dfs(r - 1, c, index + 1) or
                 dfs(r, c + 1, index + 1) or
                 dfs(r, c - 1, index + 1))
        
        # Restore
        board[r][c] = temp
        
        return found
    
    for r in range(rows):
        for c in range(cols):
            if dfs(r, c, 0):
                return True
    
    return False


# Test cases
if __name__ == "__main__":
    board = [
        ['A', 'B', 'C', 'E'],
        ['S', 'F', 'C', 'S'],
        ['A', 'D', 'E', 'E']
    ]
    
    test_words = ["ABCCED", "SEE", "ABCB"]
    
    for word in test_words:
        # Create a copy of board for each test
        board_copy = [row[:] for row in board]
        print(f"Word '{word}' exists: {exist(board_copy, word)}")
